### ISSUE TITLE
Awful sprite flickering problem

### ISSUE THREAD

**Issue by FoxFore32x** [GRRLIB/GRRLIB#57]
Hey, so i've been using GRRLIB for almost a year to make my own 2D platformer, and during all that time i've been experimenting with the library, i always had the issue that when my sprites are on top of each other (like the player on a background for example) flicker or dissapear a lot, like weirdley glichty polygons.

examples of this:

A prototype of my game (SIN-GAL) built in december 2023 for Wii;

https://x.com/FoxFore32x/status/1725726589449879637/video/2

Most recent prototype for GameCube (small epilepsy warning):

https://github.com/user-attachments/assets/d927aab7-d73d-42e3-874a-712a086c35ee

i still couldn't figure out how to draw backgrounds and object sprites without them fighting lmao

i even tried threading like in this code:

```cpp
// Define number of threads
#define NUM_THREADS 4

// Mutex for synchronizing texture loading
lwp_t threads[NUM_THREADS];
static GRRLIB_texImg* texture = NULL;
static const unsigned char* lastSpriteData = NULL;
SpriteWithCollision* globalSprite;
int imageXScale, imageYScale;
ObjectList* globalObjectList;

// Structure to pass multiple arguments to the thread function
typedef struct {
    int thread_id;
} ThreadData;

// Thread worker function
static void* draw_objects_thread(void* arg) {
    ThreadData* data = (ThreadData*)arg;  // Cast the argument
    int thread_id = data->thread_id;  // Extract thread ID

    int numObjectsPerThread = globalObjectList->numObjects / NUM_THREADS;
    int start = thread_id * numObjectsPerThread;
    int end = (thread_id == NUM_THREADS - 1) ? globalObjectList->numObjects : start + numObjectsPerThread;

    // Ensure texture is loaded once before threads
    if (lastSpriteData != globalSprite->spriteData) {
        if (texture != NULL) {
            GRRLIB_FreeTexture(texture);
        }
        texture = GRRLIB_LoadTexturePNG(globalSprite->spriteData);
        lastSpriteData = globalSprite->spriteData;
    }

    // Draw each object assigned to this thread
    for (int i = start; i < end; ++i) {
        currentObject = globalObjectList->objects[i];
        if (currentObject != NULL) {
            float screenX = (currentObject->x - currentObject->collision_mask.originX - view_camera[0].CamX);
            float screenY = (currentObject->y - currentObject->collision_mask.originY - view_camera[0].CamY);
            float objectWidth = (currentObject->width * imageXScale) * 0.03125;
            float objectHeight = (currentObject->height * imageYScale) * 0.03125;

            GRRLIB_SetHandle(texture, currentObject->collision_mask.originX, currentObject->collision_mask.originY);

            // Drawing the object
            GRRLIB_DrawImg(screenX, screenY, texture, 0, objectWidth, objectHeight, GRRLIB_WHITE);
        }
    }

    return NULL;
}

// Main draw function with threading
void draw_objects(ObjectList* obj_list, SpriteWithCollision* funnySprite, int image_xscale, int image_yscale) {
    globalSprite = funnySprite;  // Set global for threading
    globalObjectList = obj_list;
    imageXScale = image_xscale;
    imageYScale = image_yscale;

    // Load the texture before starting threads (avoid per-thread loading)
    if (lastSpriteData != globalSprite->spriteData) {
        if (texture != NULL) {
            GRRLIB_FreeTexture(texture);
        }
        texture = GRRLIB_LoadTexturePNG(globalSprite->spriteData);
        lastSpriteData = globalSprite->spriteData;
    }

    // Sort objects by depth before drawing
    sort_objects_by_depth(obj_list);

    // Create thread data for each thread
    ThreadData threadData[NUM_THREADS];

    // Create and run threads
    for (int i = 0; i < NUM_THREADS; i++) {
        threadData[i].thread_id = i;
        LWP_CreateThread(&threads[i], (void*)draw_objects_thread, &threadData[i], NULL, 0, 80);
    }

    // Wait for all threads to finish
    for (int i = 0; i < NUM_THREADS; i++) {
        LWP_JoinThread(threads[i], NULL);
    }
}
```

**Comment by Crayon2000** [comment]
Hi, I'm not really sure why you are using threads in this context? Was there a performance issue when drawing everything one after the other?

It would also be interesting to see your main game loop with the `GRRLIB_Render()` function. Is it something like:
```cpp
    while(1) {
        // Some stuff...

        draw_objects(list1, ...);
        draw_objects(list2, ...);
        draw_objects(list3, ...);

        GRRLIB_Render();  // Render the frame buffer to the TV
    }
```

**Comment by FoxFore32x** [comment]
so i managed to find a way to actually make the sprites more stable, all thanks to NewoGame / Owen.

basically the problem was that the sprites weren't being separated in the z axys, fighting with one another for which one should be displayed on top, causing the errors. Owen told me that to fix this bug basically i need to make my own "GRRLIB_DrawImg" function that includes the z axys as an input, as if i'm separting the sprites between layers for example.

here's the code:

```
extern GRRLIB_drawSettings GRRLIB_Settings;
extern Mtx                 GXmodelView2D;

static guVector axis = (guVector){0, 0, 1};

void drawing2(const f32 xpos, const f32 ypos, const f32 zpos, const GRRLIB_texImg *tex, const f32 degrees, const f32 scaleX, const f32 scaleY, const u32 color) {
    GXTexObj  texObj;
    Mtx       m, m1, m2, mv;

    if (tex == NULL || tex->data == NULL)
        return;

    GX_InitTexObj(&texObj, tex->data, tex->w, tex->h,
                  GX_TF_RGBA8, GX_CLAMP, GX_CLAMP, GX_FALSE);

    if (GRRLIB_Settings.antialias == false) {
        GX_InitTexObjLOD(&texObj, GX_NEAR, GX_NEAR,
                         0.0f, 0.0f, 0.0f, 0, 0, GX_ANISO_1);
        GX_SetCopyFilter(GX_FALSE, rmode->sample_pattern, GX_FALSE, rmode->vfilter);
    }
    else {
        GX_SetCopyFilter(rmode->aa, rmode->sample_pattern, GX_TRUE, rmode->vfilter);
    }

    GX_LoadTexObj(&texObj,      GX_TEXMAP0);
    GX_SetTevOp  (GX_TEVSTAGE0, GX_MODULATE);
    GX_SetVtxDesc(GX_VA_TEX0,   GX_DIRECT);

    guMtxIdentity  (m1);
    guMtxScaleApply(m1, m1, scaleX, scaleY, 1.0);
    guMtxRotAxisDeg(m2, &axis, degrees);
    guMtxConcat    (m2, m1, m);

    const f32 width  = tex->w * 0.5;
    const f32 height = tex->h * 0.5;

    guMtxTransApply(m, m,
        xpos +width  +tex->handlex
            -tex->offsetx +( scaleX *(-tex->handley *sin(-DegToRad(degrees))
                                      -tex->handlex *cos(-DegToRad(degrees))) ),
        ypos +height +tex->handley
            -tex->offsety +( scaleY *(-tex->handley *cos(-DegToRad(degrees))
                                      +tex->handlex *sin(-DegToRad(degrees))) ),
        zpos);
    guMtxConcat(GXmodelView2D, m, mv);

    GX_LoadPosMtxImm(mv, GX_PNMTX0);
    GX_Begin(GX_QUADS, GX_VTXFMT0, 4);
        GX_Position3f32(-width, -height, 0);
        GX_Color1u32   (color);
        GX_TexCoord2f32(0, 0);

        GX_Position3f32(width, -height, 0);
        GX_Color1u32   (color);
        GX_TexCoord2f32(1, 0);

        GX_Position3f32(width, height, 0);
        GX_Color1u32   (color);
        GX_TexCoord2f32(1, 1);

        GX_Position3f32(-width, height, 0);
        GX_Color1u32   (color);
        GX_TexCoord2f32(0, 1);
    GX_End();
    GX_LoadPosMtxImm(GXmodelView2D, GX_PNMTX0);

    GX_SetTevOp  (GX_TEVSTAGE0, GX_PASSCLR);
    GX_SetVtxDesc(GX_VA_TEX0,   GX_NONE);
}
```

basically you input a number in the "zpos" input so the sprites separate and work without issues.

Here's a demo of it (the sprites look messed up but that's more of my code being a bit buggy at the moment)


https://github.com/user-attachments/assets/159db6df-5d29-43ce-830c-fba721233f15

One thing tho, transparency is a bit of mess now, since the transparent pixels also draw over other objects


https://github.com/user-attachments/assets/97a6dc9c-9430-4d58-bfab-64ba5f0db109

**Comment by Crayon2000** [comment]
I'm closing this "issue". If you need help, you may go to the [discussions page](https://github.com/GRRLIB/GRRLIB/discussions).
